File: src/main/java/alfheim/client/render/entity/RenderEntityFlugel.java
package alfheim.client.render.entity;

import static org.lwjgl.opengl.GL11.*;

import alexsocol.asjlib.render.RenderPostShaders;
import alexsocol.asjlib.render.ShadedObject;
import alfheim.api.lib.LibResourceLocations;
import net.minecraft.client.model.ModelBase;
import net.minecraft.client.renderer.OpenGlHelper;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.entity.RenderLiving;
import net.minecraft.entity.Entity;
import net.minecraft.util.ResourceLocation;
import vazkii.botania.client.core.helper.ShaderHelper;

public class RenderEntityFlugel extends RenderLiving {

	public static final ShadedObject so = new ShadedObject(ShaderHelper.halo, RenderPostShaders.getNextAvailableRenderObjectMaterialID(), LibResourceLocations.halo) {
		
		@Override
		public void preRender() {
			glEnable(GL_BLEND);
			glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
			glShadeModel(GL_SMOOTH);
			OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, 240, 240);
			glDisable(GL_LIGHTING);
			glDisable(GL_CULL_FACE);
			glColor4f(1F, 1F, 1F, 1F);
		}
		
		@Override
		public void drawMesh() {
			Tessellator tes = Tessellator.instance;
			tes.startDrawingQuads();
			tes.addVertexWithUV(-0.75, 0, -0.75, 0, 0);
			tes.addVertexWithUV(-0.75, 0, 0.75, 0, 1);
			tes.addVertexWithUV(0.75, 0, 0.75, 1, 1);
			tes.addVertexWithUV(0.75, 0, -0.75, 1, 0);
			tes.draw();
		}
		
		@Override
		public void postRender() {
			glEnable(GL_CULL_FACE);
			//glEnable(GL_LIGHTING); breaks some other stuuf, urgh -_-
			glShadeModel(GL_FLAT);
			glDisable(GL_BLEND);
		}
	};
	
	public RenderEntityFlugel(ModelBase model, float shadowSize) {
		super(model, shadowSize);
		RenderPostShaders.registerShadedObject(so);
	}

	@Override
	public ResourceLocation getEntityTexture(Entity par1Entity) {
		return LibResourceLocations.jibril;
	}
}

File: src/main/java/alfheim/client/render/entity/RenderEntityFlugel.kt
package alfheim.client.render.entity

import alexsocol.asjlib.math.Vector3
import alexsocol.asjlib.render.*
import alfheim.api.lib.LibResourceLocations
import alfheim.client.model.entity.ModelEntityFlugel
import alfheim.common.entity.boss.EntityFlugel
import net.minecraft.client.renderer.*
import net.minecraft.client.renderer.entity.RenderLiving
import net.minecraft.entity.Entity
import net.minecraft.util.ResourceLocation
import org.lwjgl.opengl.GL11
import vazkii.botania.client.core.handler.BossBarHandler
import vazkii.botania.client.core.helper.ShaderHelper

object RenderEntityFlugel: RenderLiving(ModelEntityFlugel(), 0.25f) {
	
	override fun getEntityTexture(entity: Entity?) =
		if (entity is EntityFlugel) getEntityTexture(entity) else LibResourceLocations.jibril
	
	fun getEntityTexture(flugel: EntityFlugel): ResourceLocation {
		if (Vector3.fromEntity(flugel) != Vector3.zero) BossBarHandler.setCurrentBoss(flugel)
		
		return if (flugel.isUltraMode) LibResourceLocations.jibrilDark else LibResourceLocations.jibril
	}
	
	val so: ShadedObject = object: ShadedObject(ShaderHelper.halo, RenderPostShaders.nextAvailableRenderObjectMaterialID, LibResourceLocations.halo) {
		
		var lastX = 0f
		var lastY = 0f
		
		override fun preRender() {
			GL11.glEnable(GL11.GL_BLEND)
			GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)
			GL11.glShadeModel(GL11.GL_SMOOTH)
			
			lastX = OpenGlHelper.lastBrightnessX
			lastY = OpenGlHelper.lastBrightnessY
			
			OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, 240f, 240f)
			GL11.glDisable(GL11.GL_LIGHTING)
			GL11.glDisable(GL11.GL_CULL_FACE)
			GL11.glColor4f(1f, 1f, 1f, 1f)
		}
		
		override fun drawMesh() {
			val tes = Tessellator.instance
			tes.startDrawingQuads()
			tes.addVertexWithUV(-0.75, 0.0, -0.75, 0.0, 0.0)
			tes.addVertexWithUV(-0.75, 0.0, 0.75, 0.0, 1.0)
			tes.addVertexWithUV(0.75, 0.0, 0.75, 1.0, 1.0)
			tes.addVertexWithUV(0.75, 0.0, -0.75, 1.0, 0.0)
			tes.draw()
		}
		
		override fun postRender() {
			GL11.glEnable(GL11.GL_CULL_FACE)
			//glEnable(GL_LIGHTING); breaks some other stuff, urgh -_-
			GL11.glShadeModel(GL11.GL_FLAT)
			GL11.glDisable(GL11.GL_BLEND)
			
			OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, lastX, lastY)
		}
	}.also { RenderPostShaders.registerShadedObject(it) }
}